### ISSUE TITLE
OSGI compatibility

### ISSUE THREAD

**Issue by agjini** [neo4j/neo4j-java-driver#537]
### Shaded jar failed to run as an OSGI bundle

We try to run neo4j-java-driver as an OSGI bundle into an OSGI container (felix).

We also provides required dependencies for neo4j driver to works : `io.netty` and `org.hdrhistogram` 
as external jars.

When we launch the application, one of our service try to instantiate the neo4j driver into the activate() method by calling `GraphDatabase.driver` method.

```java
    @Activate
    public void activate(Neo4jClientConfig config) {
        String host = config.neo4j_host();
        int boltPort = config.bolt_port();

        String hostHttp = config.neo4j_host();
        int portHttp = config.neo4j_port();
        String protocolHttp = config.neo4j_protocol();
        init(URI.create(protocolHttp + "://" + hostHttp + ":" + portHttp), URI.create("bolt://" + host + ':' + boltPort));
    }

    public void init(URI httpUri, URI boltUri) {
        this.host = new HttpHost(httpUri.getHost(), httpUri.getPort(), httpUri.getScheme());
        this.client = new HttpClientBuilder().build();
        driver = GraphDatabase.driver(boltUri, Config.build().withoutEncryption().toConfig());
    }
```

We use the `neo4j-java-driver:1.6.3` which embeds netty and hdrhistogram dependencies with rewritten package names (shaded jar).
When we use this bundle, the `GraphDatabase.driver()` method failed with the following error : 

```log
16:30:54.585 [ERROR]     o.o.u.l.i.o.LoggerService {FelixStartLevel} - [neo4j-bolt-client(73)] The activate method has thrown an exception
java.lang.NoClassDefFoundError: Could not initialize class org.neo4j.driver.internal.shaded.io.netty.channel.DefaultChannelId                 
        at org.neo4j.driver.internal.shaded.io.netty.channel.AbstractChannel.newId(AbstractChannel.java:111)              
        at org.neo4j.driver.internal.shaded.io.netty.channel.AbstractChannel.<init>(AbstractChannel.java:83)                            
        at org.neo4j.driver.internal.shaded.io.netty.bootstrap.FailedChannel.<init>(FailedChannel.java:33)                       
        at org.neo4j.driver.internal.shaded.io.netty.bootstrap.AbstractBootstrap.initAndRegister(AbstractBootstrap.java:330)
        at org.neo4j.driver.internal.shaded.io.netty.bootstrap.Bootstrap.doResolveAndConnect(Bootstrap.java:163)                                         
        at org.neo4j.driver.internal.shaded.io.netty.bootstrap.Bootstrap.connect(Bootstrap.java:145)                                                                       
        at org.neo4j.driver.internal.async.ChannelConnectorImpl.connect(ChannelConnectorImpl.java:78)                                                    
        at org.neo4j.driver.internal.async.pool.NettyChannelPool.connectChannel(NettyChannelPool.java:64)                                                                                                          
        at org.neo4j.driver.internal.shaded.io.netty.channel.pool.SimpleChannelPool.acquireHealthyFromPoolOrNew(SimpleChannelPool.java:179)
        at org.neo4j.driver.internal.shaded.io.netty.channel.pool.SimpleChannelPool.acquire(SimpleChannelPool.java:164)                                                                                            
        at org.neo4j.driver.internal.shaded.io.netty.channel.pool.FixedChannelPool.acquire0(FixedChannelPool.java:266)
        at org.neo4j.driver.internal.shaded.io.netty.channel.pool.FixedChannelPool.access$300(FixedChannelPool.java:38)            
        at org.neo4j.driver.internal.shaded.io.netty.channel.pool.FixedChannelPool$3.run(FixedChannelPool.java:240)                       
        at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
        at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:404)
        at org.neo4j.driver.internal.shaded.io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)                         
        at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:886)
        at org.neo4j.driver.internal.shaded.io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(Thread.java:748)
```

### Fixing the issue

We successfully fix this issue by deploying the non shaded jar i.e. the original neo4j-driver-java bundle.

### 

Is there any reason why the neo4j-java-driver now embeds these dependencies into the jar (with rewritten packages), it seems to be the case since versions > 1.4.x ?

It seems that packaging a simple jar is the solution for OSGI compatibility anyway.

Thanks for your help

**Comment by lutovich** [comment]
Hi @agjini,

Driver embeds those dependencies in order to not create dependency version conflicts for the user code. This is valuable when user code depends on a different version of netty or hdrhistogram than the driver.

Could you please provide an example OSGI project where reported failure can be observed?

**Comment by lutovich** [comment]
Ping @agjini. Any updates on this? We do not have much experience with OSGI and it would be great to have a reproducer project for this issue.

**Comment by agjini** [comment]
Hi and thanks for the reply.
Yes I will provide an example project as soon as I have a moment to package a reproducer project based on osgi (as tiny as possible).
I'll try to find a moment soon.

**Comment by azzazzel** [comment]
I just had the same issue. I'm still digging into it as it seams very weird at first. 

However I noticed a more generic problem.  The MANIFEST.MF file contains imports for `io.netty.*` which makes it unable to resolve unless one provides `netty` and  `hdrhistogram` bundles to OSGi's resolver as well. However in reality those seem to not be needed as what the drives internally uses is the shaded versions of those.

**Comment by lutovich** [comment]
@azzazzel do you have an example project we can use to reproduce the problem?

**Comment by azzazzel** [comment]
I think I can. May be this weekend. 

Meanwhile I did some investigation and I think what is happening is that an exception is being thrown in the static block of `DefaultChannelId`: https://github.com/netty/netty/blob/netty-4.1.22.Final/transport/src/main/java/io/netty/channel/DefaultChannelId.java#L58-L106

**Comment by agjini** [comment]
This is the same issue I've experimented. The reason is the same.
You are right the other solution might be to remove `netty` and `hdrhistogram` dependencies from `Import-Package` in `MANIFEST.MF` and use the shaded jar (which depends on internal rewritten packages of netty and histogram). This should work, but the Jar size will be heavy.

**Comment by azzazzel** [comment]
@agjini actually this is another problem and not a solution. The `netty` and `hdrhistogram` bundles are not used at runtime at all even now. They are just required for OSGi resolver to do the wiring but it's useless wiring as the code uses the shaded packages anyways. I plan to submit a PR fixing that issue soon but it'll not solve the issue we discuss here.

**Comment by azzazzel** [comment]
So quick workaround is to 
 - download the driver source code
 - remove (comment out) the shade plugin in pom file
 - build and install the project 
 - use that jar file with `netty` and `hdrhistogram` bundles at runtime 

Everything resolves and works just fine without the shading. The shading causes all sorts of problems but most notably:
  - integration tests do not test what is actually being shipped (see https://github.com/neo4j/neo4j-java-driver/blob/1.7/driver/pom.xml#L201-L211)
  - source jar does not contain the shaded classes which makes it impossible to debug the code
  - there is a chicken and egg problem between generating proper metadata and shading
 
I was hoping to have an integration test for this issue by now but due to all those issues it's almost impossible to write one ;(

**Comment by lutovich** [comment]
@azzazzel shading seems to only be a problem for OSGI environment. It is great for everyone else for reasons mentioned above and in this [SO question](https://stackoverflow.com/questions/13620281/what-is-the-maven-shade-plugin-used-for-and-why-would-you-want-to-relocate-java).

Please feel free to submit a PR with fix or provide a reproducer project. We do not have much experience with OSGI.

**Comment by azzazzel** [comment]
@lutovich I get what you are saying. And even though it's not only OSGi _(my guess is the problem exists in any non-flat classloader environment such as JBoss modules or JPMS with layers)_, it's not my goal to convince you to change that. 

I'm trying to give you a test that can reproduce the bug and a potential solution but shading makes it almost impossible to test and debug. Even your current integration tests do not test the actual shaded result.

**Comment by azzazzel** [comment]
@lutovich would you consider delivering shaded and non-shaded versions of the driver?

**Comment by agjini** [comment]
Thanks for your help guys.
I think the solution to deliver the non shaded version of the jar is a good solution. This is the solution I've tested with success in my OSGI app.
If the shaded version is useful to some situations (to be able to deploy two version of netty jar in Java app) it has to be deployed too but maybe with a meaningfull classifier.

Is it still necessary that I'll provide you a reproducer project (I'm closed to have a mini project but it is not running yet) ?

**Comment by azzazzel** [comment]
Hey @agjini I think I made the shaded bundle work in OSGi. Can you test the PR above and confirm it works for you too?

**Comment by lutovich** [comment]
@agjini could you please test the fix provided in https://github.com/neo4j/neo4j-java-driver/pull/545? Reproducer project would still be valuable, we could include it as an integration test in the build pipeline.

@azzazzel why is a non-shaded jar needed when the problem is fixed by the PR?

**Comment by azzazzel** [comment]
@lutovich it is not strictly needed but shading introduces at least 2 (IMHO serious) issues:
 - it's poor design and violates coherence rules. OSGi is the first to show you that, but you would have the same issue if you ever want to move to JPMS (Jigsaw). In fact JPMS explicitly forbids two modules to contain the same package, so the JVM will not even start if the driver is placed on a module path with `netty` or `hdrhistogram` modules.
 - it's a maintenance issue. Again it's the OSGi that demonstrates it but it will be the same with upcoming tools for other module systems. The tools typically need to run before `package` phase to prepare what goes in the package (`MANIFEST.MF` for OSGi, `module-info.class` for JPMS, ...). However later on shading changes their source of truth and renders them invalid / inaccurate.  Therefore one has to manually tell the tools (which is what the fix here does) to produce metadata that will be valid after the shading changes. For example, the moment you update the `netty` or `hdrhistogram` version this fix may no longer work. You'll have to examine both and make sure you properly transfer their metadata to the shaded driver.

**Comment by lutovich** [comment]
@azzazzel question about your first point. Right now shading relocates classes to a package prefixed with `org.neo4j.driver.internal.shaded`. Doesn't this prevent split packages?

**Comment by agjini** [comment]
Yes it prevent split package.
Since the shaded Jar creation, relocate the package to
'org.neo4j.driver.internal.shaded',
the Java code (when crea
OSGI doesn't allow

**Comment by azzazzel** [comment]
@lutovich my bad - forgot about the changed package names. You are right - that would work and the JPMS will likely wire properly. However when `netty` or `hdrhistogram` become JPMS modules they may require some other modules. You'll have to transfer those requirements to the shaded `neo4j-java-driver` module for JPMS to allow you to access those 3rd party packages.

**Comment by agjini** [comment]
@lutovich I've successfully tested the pr/545 made by @azzazzel : it launches well in our OSGI project.
I would prefer not to have a shaded jar (around 5 Mb) but it works well !

**Comment by lutovich** [comment]
PR with the fix is now merged, thanks @azzazzel! 1.7.1 version will contain the fix and should be released in a week or at most two.

@agjini reproducer project is still super welcome if you have time.

Shading will remain for 1.x series of the driver. We will consider the suggested packaging changes for the next major release. Publishing two different artifacts or using classifiers could be an option. JPMS and modularization require a JDK version bump and could be problematic because of [this issue](https://issues.apache.org/jira/browse/MSHADE-234).

**Comment by lutovich** [comment]
Driver 1.7.1 containing the given fix is now released.
